usererrors.go

```
@@ -0,0 +1,162 @@
+		err.successfulKeyGroups())
+}
+
+func (err *getDataKeyError) UserError() string {
+	var groupErrs []string
+	for i, res := range err.GroupResults {
+		groupErr := decryptGroupError{
+			err:       res,
+			groupName: fmt.Sprintf("%d", i),
+		}
+		groupErrs = append(groupErrs, groupErr.UserError())
+	}
+	var trailer string
+	if err.RequiredSuccessfulKeyGroups == 0 {
+		trailer = "Recovery failed because no master key was able to decrypt " +
+			"the file. In order for SOPS to recover the file, at least one key " +
+			"has to be successful, but none were."
+	} else {
+		trailer = fmt.Sprintf("Recovery failed because the file was "+
+			"encrypted with a Shamir threshold of %d, but only %d part(s) "+
+			"were successfully recovered, one for each successful key group. "+
+			"In order for SOPS to recover the file, at least %d groups have "+
+			"to be successful. In order for a group to be successful, "+
+			"decryption has to succeed with any of the keys in that key group.",
+			err.RequiredSuccessfulKeyGroups, err.successfulKeyGroups(),
+			err.RequiredSuccessfulKeyGroups)
+	}
+	trailer = wordwrap.WrapString(trailer, 75)
+	return fmt.Sprintf("Failed to get the data key required to "+
+		"decrypt the SOPS file.\n\n%s\n\n%s",
+		strings.Join(groupErrs, "\n\n"), trailer)
+}
+
+type decryptGroupError struct {
+	groupName string
+	err       error
+}
+
+func (r *decryptGroupError) Error() string {
+	return fmt.Sprintf("could not decryt group %s: %s", r.groupName, r.err)
+}
+
+func (r *decryptGroupError) UserError() string {
+	var status string
+	if r.err == nil {
+		status = statusSuccess
+	} else {
+		status = statusFailed
+	}
+	header := fmt.Sprintf(`Group %s: %s`, r.groupName, status)
+	if r.err == nil {
+		return header
+	}
+	message := r.err.Error()
+	if userError, ok := r.err.(UserError); ok {
+		message = userError.UserError()
+	}
+	reader := prefixer.New(strings.NewReader(message), "  ")
+	errMsg, _ := ioutil.ReadAll(reader)
```

maybe shouldn't ignore the error here?